### ISSUE TITLE
High docker cpu usage with Nomad on an idle node

### ISSUE THREAD

**Issue by drscre** [hashicorp/nomad#3662]
**TL/DR** Streaming stats from docker containers to nomad-executor seems to cause high system cpu usage by docker, probably due to lock contention. The more docker containers on the node are running, the more noticable this effect is.

Nomad v0.7.0

We are investigating an issue where Nomad is causing strange cpu load. 
We observe it both on our production (amazon ec2 ubuntu 14.04) nodes and our development node with different Nomad and docker versions.

We see significant system/user cpu usage and strange periodic load average spikes even on development node which has no load at all.

In the graph below you can see cpu usage of an idle node running Nomad with ~40 idle docker containers with no load.
~12% of user cpu and 12% of system cpu is used.

`top` shows that these 12% cpu are used by `docker-containerd`, so it is natural to blame docker.

But when we stop Nomad while keeping containers running, the abnormal load disappears.

Also, there is a significant amount of *system* cpu usage and high la spikes. So i presume this has something to do with linux futexes, as there is no io/networking load at all.

Can this be a bug in a way Nomad interacts with docker containers?
Can we somehow disable this interaction to debug the problem?

![nomad_cpu_load](https://user-images.githubusercontent.com/844519/34017299-438480d4-e136-11e7-9ddb-d3a105a1a90a.png)

During la spikes `strace`ing `nomad-executor`s shows a lot of `futex` calls:
```
.... same calls
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
pselect6(0, NULL, NULL, NULL, {0, 3000}, NULL) = 0 (Timeout)
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = -1 EAGAIN (Resource temporarily unavailable)
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = -1 EAGAIN (Resource temporarily unavailable)
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = -1 EAGAIN (Resource temporarily unavailable)
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = -1 EAGAIN (Resource temporarily unavailable)
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = 0
futex(0x206da10, FUTEX_WAIT, 0, NULL)   = -1 EAGAIN (Resource temporarily unavailable)
... same calls
```

**Comment by dadgar** [comment]
@drscre After starting a container, the only thing Nomad will do is collect stats. I would try increasing this: https://www.nomadproject.io/docs/agent/configuration/telemetry.html#collection_interval

Would love to hear back the results.

**Comment by drscre** [comment]
Increasing collection interval to 1hour was the first thing i tried. But it did not help.

What does nomad-executor process do? Why does it need all those futex calls?

**Comment by dadgar** [comment]
@drscre When you said you killed Nomad did you kill the executors too?

**Comment by drscre** [comment]
Yes, just left docker containers running

**Comment by dadgar** [comment]
@drscre Are the docker containers logging heavily?

**Comment by drscre** [comment]
No, the system is completely idle, doing nothing. No requests, no load.

**Comment by drscre** [comment]
@dadgar 
I tried to recompile Nomad without stats collection 
I commented out these lines:
https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L818
https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L1596

And it looks like it helped , though not entirely:
![screenshot-2017-12-15-04-07-1513300023](https://user-images.githubusercontent.com/844519/34021983-ad88fcb4-e14d-11e7-9df1-1fe00b4f5c63.png)

**Comment by drscre** [comment]
Also, I don't see `StatsCollectionInterval` config option being used in `collectStats()` function. 
It looks more like it opens a long-living connection to docker who constantly streams back resource usage.

**Comment by drscre** [comment]
@dadgar You can easily reproduce this issue using getting started guide: https://www.nomadproject.io/intro/getting-started/jobs.html
Just launch 40 copies of example redis job and observe cpu usage with `top`:
![cpu_load](https://user-images.githubusercontent.com/844519/34079699-b675c6ca-e343-11e7-9802-e628792bce38.png)


Then kill nomad and executors with -SIGKILL leaving docker containers running and observe cpu usage again:
![no_cpu_load](https://user-images.githubusercontent.com/844519/34079706-c5b6da3e-e343-11e7-92b7-c5b8fbeabb8e.png)

**Comment by hsmade** [comment]
> Also, I don't see StatsCollectionInterval config option being used in DockerHandle::collectStats() function.
It looks more like it opens a long-living connection to docker who constantly streams back resource usage.

https://github.com/hashicorp/nomad/blob/v0.7.1/client/client.go#L1940 (0.7.1)
Timer is first set at 0 / now. Then on each iteration, it is set for `c.config.StatsCollectionInterval`. So it is actually used.

**Comment by drscre** [comment]
@hsmade Yes, this is not container stats collection.

So far it looks like this is not a Nomad but rather a Docker bug. It occurs when you stream stats from many containers at the same times.
It is strange that nobody at Hashicorp has observed this extra cpu consumption so far, because it noticeable even when with a small amount of containers.

As a workaround, it would be great to have an option to collect container stats periodically instead of keeping long-living streaming connections to docker.

**Comment by hsmade** [comment]
Can you reproduce this with `docker stats`? Because that does the same thing, it listens for stats on the docker api.

**Comment by hsmade** [comment]
Also, do you have a vagrant file? I'm trying to reproduce your example (as we might see the same in our prod env). I have a debian/stretch with 2G/2 cores, running on a macbook pro, and I don't get the cpu load you get:

    top - 14:07:58 up 4 min,  1 user,  load average: 3.19, 1.58, 0.65
    Tasks: 319 total,   1 running, 318 sleeping,   0 stopped,   0 zombie
    %Cpu(s):  3.2 us,  4.4 sy,  0.0 ni, 92.1 id,  0.0 wa,  0.0 hi,  0.3 si,  0.0 st

**Comment by hsmade** [comment]
Maybe the problem here is that docker is working too hard to get the stats, and doesn't allow to set the interval on the API. If it would, nomad could implement that and not spend so much time on gathering the stats as well.

**Comment by drscre** [comment]
From what i see, docker stats API does not provide any collection interval customization.
It seems that a possible alternative for Nomad could be to query docker for stats periodically in non-streaming mode.
I.e. send `GET /v1.21/containers/redis1/stats?stream=0` instead of keeping streaming connection.

**Comment by dropje86** [comment]
FWIW, here's a stack dump of the nomad-executor showing the aforementioned behavior (lots and lots of futex syscalls): https://pastebin.com/QKgW0W9Q

One thing that stood out to me is the fact that the executor runs the Syslog server. So just for the sake of it I then tried switching a single job to use fluentd as a logging driver. So far the executor running that job behaves normally while others (running the default syslog) are performing huge amounts of syscalls.

**Comment by hvindin** [comment]
We're seeing this on nomad v0.8.1currently.

Its gotten to the point where we have needed to arbitrarily increase the resource profile of jobs so that we don't get more than ~10 jobs running (mostly idle java and node API's in containers) per host (average host size is 4xvCPU's@2GHz + 16G RAM) otherwise we see unpredictable and very significant spikea in CPU usage to the extent that containers start to die and restart, triggering something of a death spiral in the cluster.

From what I can find the underlying cause ie almost definitely a containerd problem, but it feels like nomad should be able to mitigate this given the sinificant perfornance overhead it introduces.

FWIW i tried the "change the logging driver to fluentd" thing mentioned by @dropje86 but this didn't significantly improve the CPU situation.

**Comment by drscre** [comment]
@hvindin Are you sure it's not the jobs themselves who are eating CPU? Do you monitor CPU usage of jobs?

The case you are describing seems a bit different from what we experience. In our case there are spikes of *load average* and some extra constant *system* cpu usage, but nothing that critical.

The reason for this extra system CPU usage is that Nomad is using containerd streaming API to collect resource usage from docker containers. And this containerd resource streaming API method seems to be poorly implemented and eats extra CPU. The more containers are running, the more noticable the extra system CPU usage is.

But in your case with only 10 containers running this should hardly be an issue.
